core/src/main/scala/epollcat/unsafe/KqueueExecutorScheduler.scala

```diff
@@ -65,53 +65,54 @@ private[unsafe] final class KqueueExecutorScheduler(
         }
       }
 
-      val timeoutSpec =
-        if (timeoutIsInfinite || timeoutIsZero) null
-        else {
-          val ts = stackalloc[time.timespec]()
-          val sec = timeout.toSeconds
-          ts.tv_sec = sec
-          ts.tv_nsec = (timeout - sec.seconds).toNanos
-          ts
-        }
+      if (!noCallbacks) {
+        val timeoutSpec =
+          if (timeoutIsInfinite || timeoutIsZero) null
+          else {
```

Comment:
What happens if there are no callbacks, but there is `0 < timeout < infinity`? In that case `poll` should still sleep for the requested time. This is not necessary for correctness, but it is important for performance (if we don't sleep, we are busy-looping). That is why the test suite currently doesn't catch this.

Reply:
Thank you for the timely review and tag-team development. 

re: `0 < timeout < infinity` & busy looping

Isn't prevention of busy looping the responsibility of the caller?  I do not know 
that scheduler code.  What is the usually timeout?  Zero, something else.

The hang that I saw was large for a scheduler.  I know I let it run two or
three minutes whilst I did code fixes in another window. In a debug 
branch I can put some code in to see if it is 0 (which can mean either
return immediately (do not wait), or wait-forever) or a largish number.

I thought that poll() was supposed to minimize wait and return to 
the caller, so the caller could schedule something else.

Reply:
Yes, perhaps the documented contract for `poll` needs to be clarified.

> I thought that poll() was supposed to minimize wait and return to
the caller, so the caller could schedule something else.

When the caller invokes `poll(...)` with a timeout of `x` seconds, it means that the caller has no work scheduled for `x` seconds and thus it is fully delegating to `poll(...)`. If `poll` returns before `timeout` (and no new work was scheduled due to an I/O event), then the caller will simply invoke `poll(...)` again for the remaining time.

Reply:
The event-loop algorithm is implemented here.
https://github.com/typelevel/cats-effect/blob/7ca03db50342773a79a01ecf137d953408ac6b1d/core/native/src/main/scala/cats/effect/unsafe/PollingExecutorScheduler.scala#L77-L123

Reply:
That info helps.    The poll is being called with INFINITE timeout, 2 events change, and 1 callback.
There was a open & close which came and went before poll() got called. I suspect the math
is not working out and an extra event-change is sneaking in.

I am going to move this to WIP,  do more tracing, and read the URL you posted.
I'll post the provoking code in the issue.

Reply:
You are right. I believe your general approach is correct, we just need to adjust the conditional to account for the case when `0 < timeout < infinity` in which case it should proceed with the `kevent64()` call.